src/Compilers/CSharp/Test/Semantic/Semantics/UsingStatementTests.cs

```diff
@@ -1452,6 +1453,280 @@ static void Main()
+                );
+        }
+
+        [Fact]
+        public void UsingPatternAsyncAndSyncTest()
+        {
+            var source = @"
+namespace System
+{
+    public interface IAsyncDisposable
+    {
+        System.Threading.Tasks.ValueTask DisposeAsync();
+    }
+}
+class C1
+{
+    public C1() { }
+    public System.Threading.Tasks.ValueTask DisposeAsync() { System.Console.Write(""Dispose async"");return new System.Threading.Tasks.ValueTask(System.Threading.Tasks.Task.CompletedTask); }
+    public void Dispose() { System.Console.Write(""Dispose; ""); }
+}
+
+class C2
+{
+    static async System.Threading.Tasks.Task Main()
+    {
+        using (C1 c = new C1())
+        {
+        }
+        await using (C1 c = new C1())
+        {
+        }
+    }
+}";
+            var compilation = CreateCompilationWithTasksExtensions(source, options: TestOptions.DebugExe).VerifyDiagnostics();
+
+            CompileAndVerify(compilation, expectedOutput: "Dispose; Dispose async");
+        }
+
+        [Fact]
+        public void UsingPatternAsyncExtensionMethodTest()
+        {
+            var source = @"
+namespace System
+{
+    public interface IAsyncDisposable
+    {
+        System.Threading.Tasks.ValueTask DisposeAsync();
+    }
+}
+class C1
+{
+}
+
+static class C2 
+{
+    public static System.Threading.Tasks.ValueTask DisposeAsync(this C1 c1) { System.Console.WriteLine(""Dispose async""); return new System.Threading.Tasks.ValueTask(System.Threading.Tasks.Task.CompletedTask); }
+}
+
+class C3
+{
+    static async System.Threading.Tasks.Task Main()
+    {
+        await using (C1 c = new C1())
+        {
+        }
+    }
+}";
+            var compilation = CreateCompilationWithTasksExtensions(source, options: TestOptions.DebugExe).VerifyDiagnostics();
+
+            CompileAndVerify(compilation, expectedOutput: "Dispose async");
+        }
+
+        [Fact]
+        public void UsingPatternAsyncWrongReturnTypeTest()
+        {
+            var source = @"
+namespace System
+{
+    public interface IAsyncDisposable
+    {
+        System.Threading.Tasks.ValueTask DisposeAsync();
+    }
+}
+class C1
+{
+    public bool DisposeAsync() { return false; }
+}
+
+class C2
+{
+    static async System.Threading.Tasks.Task Main()
+    {
+        await using (C1 c = new C1())
+        {
+        }
+    }
+}";
+            CreateCompilationWithTasksExtensions(source).VerifyDiagnostics(
+                // (18,22): warning CS0280: 'C1' does not implement the 'disposable' pattern. 'C1.DisposeAsync()' has the wrong signature.
+                //         await using (C1 c = new C1())
+                Diagnostic(ErrorCode.WRN_PatternBadSignature, "C1 c = new C1()").WithArguments("C1", "disposable", "C1.DisposeAsync()").WithLocation(18, 22),
+                // (18,22): error CS8410: 'C1': type used in an async using statement must be implicitly convertible to 'System.IAsyncDisposable'
```

Comment:
>async [](start = 64, length = 5)

async -> asynchronous, to avoid implying an `async using` syntax

Reply:
Was this addressed?

---
In reply to: [234318277](https://github.com/dotnet/roslyn/pull/31201#discussion_r234318277) [](ancestors = 234318277)